### ISSUE TITLE
Implement shallow routing

### ISSUE THREAD

**Issue by hlynurl** [nuxt/nuxt.js#6147]
### What problem does this feature solve?

I want a clean url scheme which goes something like example.com/:company/:products/:id.

I change the routes by clicking on a tree like list in my sidebar. I already have all the data from the first load so I want to change the url without re-rendering the whole page.

This is possible in Next js with something called ["shallow routing"](https://github.com/zeit/next.js#shallow-routing)

<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c9569">#c9569</a>)</em></sub></div>

**Comment by pimlie** [comment]
Try to use [`<keep-alive>`](https://vuejs.org/v2/api/#keep-alive) on your [nuxt-child](https://nuxtjs.org/api/components-nuxt-child#the-lt-nuxt-child-gt-component)

**Comment by hlynurl** [comment]
Thanks @pimlie I was able to make it work that way 👍

**Comment by niksajanjic** [comment]
Hi, I'd like to avoid creating a new issue, but I believe the "Implement shallow routing" should be revisited/reopened. I'll explain it thoroughly.

There is an example of using `keep-alive` here in Nuxt repo that does not work at the moment (nuxt 2.13.3):
https://github.com/nuxt/nuxt.js/tree/dev/examples/with-keep-alive

If we try to run this example, we will see that `asyncData` inside `pages/about` is being called every time we land on `/about` page. We can `console.log` message inside to prove that. What does work is if we change `asyncData` for `fetch`. In this case, it will be called only for the first time and never again (until we refresh the app).

I guess at some point `asyncData` was cached, but with the latest version of Nuxt (2.13.3) it does not. Therefore, this example should either be changed or removed.

Furthermore, as this issue proposes, Nuxt should/could support `shallow routing`, because currently, it seems like it doesn't. Next.js has this feature (https://nextjs.org/docs/routing/shallow-routing):

```Shallow routing allows you to change the URL without running data fetching methods again, that includes getServerSideProps, getStaticProps, and getInitialProps.```

Basically, any method that fetches data will get cached when using shallow routing. Also, shallow routing works by adding `shallow` as a param to link component, to a `router.push` method or watch for query changes inside the component to decide whether we want to fetch new data or not. As we can see, there is a lot of ways to implement `shallow routing`.

The only way I could prevent `asyncData` from fetching data again in Nuxt was by following answer in this issue:
https://github.com/nuxt/nuxt.js/issues/1322

So, basically, I have this code:
```JS
async asyncData({ $axios, from }) {
  if (from && from.matched[0].name === 'posts-id') return

  const someData = await $axios.$get(`${process.env.API_URL}/someEndpoint`)
  return {
    someData
  }
}
```

This code fetches data only once and it won't fetch data again. My component has the data from the previous fetch so it works on route change. Unfortunately, this breaks `hot reloading`. While this would work in production, if I change anything in this file in dev mode it will break and show `404` page. The reason is that component loses its data from the first fetch on hot reload and it won't make another fetch because of this `if` statement above that prevents it. And then if your template expects that data it will just break, show 404 page and error in console.

Even if there is a way to make `if` statement better and include logic for `hot reloading` I still feel there should be a much better way to implement `shallow routing`. Also, it should work both for `asyncData` and `fetch`, because I don't feel devs should be forced to use `fetch` method, they should be able to take advantage of SSR (similar like Next.js does). Another problem with `fetch` method is that devs have additional overhead with spinners and handling errors themselves.

Lastly, one powerful feature Nuxt.js provides is by deciding when to shallow route and when not to by passing prop to link:
```js
<Link href="/posts" shallow>
  <a>Posts</a>
</Link>
```
This is awesome because we can decide when we want to fetch posts again and when not to by just clicking a specific link. This gives us opportunity to decide on shallow routing depending from which part of the app are we coming from. I had this exact situation with `/:surveyId/:questionId` where I get all the data with `surveyId`, but then each question is it's own route and changing it doesn't require new fetch as we already have data. So, question links were shallow, but new survey links were not which meant new data fetch. With Nuxt, I implemented this with a lot of unnecessary logic inside `asyncData` and I had to manually cache first fetch so I don't break `hot reloading`.
